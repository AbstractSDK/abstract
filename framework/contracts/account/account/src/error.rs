use cosmwasm_std::StdError;

use manager::error::ManagerError;
use proxy::error::ProxyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error("{0}")]
    Std(#[from] StdError),

    // #[error("{0}")]
    // Abstract(#[from] AbstractError),

    // #[error("{0}")]
    // AbstractSdk(#[from] AbstractSdkError),

    // #[error("{0}")]
    // Admin(#[from] AdminError),
    #[error(transparent)]
    Manager(#[from] ManagerError),

    #[error(transparent)]
    Proxy(#[from] ProxyError),
}
