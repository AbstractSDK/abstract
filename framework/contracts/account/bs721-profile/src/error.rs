use cosmwasm_std::StdError;
use cw_controllers::AdminError;
use cw_ownable::OwnershipError;
use cw_utils::PaymentError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ContractError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    OwnershipError(#[from] OwnershipError),
    #[error("{0}")]
    Payment(#[from] PaymentError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Base(#[from] bs721_base::ContractError),

    #[error("NameNotFound")]
    NameNotFound {},

    #[error("AddressAlreadyMapped")]
    AddressAlreadyMapped {},

    #[error("RecordNameAlreadyExists")]
    RecordNameAlreadyExists {},

    #[error("RecordNameEmpty")]
    RecordNameEmpty {},

    #[error("RecordNameTooLong")]
    RecordNameTooLong {},

    #[error("RecordValueTooLong")]
    RecordValueTooLong {},

    #[error("RecordValueEmpty")]
    RecordValueEmpty {},

    #[error("UnauthorizedVerification")]
    UnauthorizedVerification {},

    #[error("Invalid Metadata")]
    InvalidMetadata {},

    #[error("Unauthorized: Not contract creator or admin")]
    UnauthorizedCreatorOrAdmin {},

    #[error("TooManyRecords max: {max}")]
    TooManyRecords { max: u32 },

    #[error("NotImplemented")]
    NotImplemented {},
}
