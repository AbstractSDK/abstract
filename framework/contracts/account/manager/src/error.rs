use abstract_core::{
    objects::{validation::ValidationError, version_control::VersionControlError},
    AbstractError,
};
use abstract_sdk::{core::objects::module::ModuleInfo, AbstractSdkError};
use cosmwasm_std::{Instantiate2AddressError, StdError};
use cw_controllers::AdminError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ManagerError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Validation(#[from] ValidationError),

    #[error("{0}")]
    Ownership(#[from] cw_ownable::OwnershipError),

    #[error("{0}")]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error("{0}")]
    VersionControlError(#[from] VersionControlError),

    #[error("Module with id: {0} is already installed")]
    ModuleAlreadyInstalled(String),

    #[error("Cannot remove module because {0:?} depend(s) on it.")]
    ModuleHasDependents(Vec<String>),

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("The name of the proposed module can not have length 0.")]
    InvalidModuleName {},

    #[error("Registering module fails because caller is not module factory")]
    CallerNotModuleFactory {},

    #[error("A migrate msg is required when when migrating this module")]
    MsgRequired {},

    #[error("{0} not upgradable")]
    NotUpgradeable(ModuleInfo),

    #[error("Cannot migrate {} twice", module_id)]
    DuplicateModuleMigration { module_id: String },

    #[error("Your account is currently suspended")]
    AccountSuspended {},

    #[error("The provided contract version {0} is lower than the current version {1}")]
    OlderVersion(String, String),

    #[error("The provided module {0} was not found")]
    ModuleNotFound(String),

    #[error("Module {module_id} with version {version} does not fit requirement {comp}, post_migration: {post_migration}")]
    VersionRequirementNotMet {
        module_id: String,
        version: String,
        comp: String,
        post_migration: bool,
    },

    #[error("module {0} is a dependency of {1} and is not installed.")]
    DependencyNotMet(String, String),

    #[error("The provided module {0} has an invalid module reference.")]
    InvalidReference(ModuleInfo),

    #[error("Cannot remove proxy")]
    CannotRemoveProxy {},

    #[error("No updates were included")]
    NoUpdates {},

    #[error("invalid configuration action, {}", error)]
    InvalidConfigAction { error: StdError },

    #[error("Must use ProposeOwner to change owner")]
    MustUseProposeOwner {},

    #[error("The address {0} doesn't have an owner, the manager can't determine admin right")]
    NoContractOwner(String),

    #[error("
            Checking the admin recursively failed. 
            You either have the an error in your sub-account configuration or you are not authorized to make this call.
    ")]
    SubAccountAdminVerification,

    #[error("Removing sub account failed")]
    SubAccountRemovalFailed {},

    #[error("Register of sub account failed")]
    SubAccountRegisterFailed {},

    #[error("Can't renounce account, with active sub account")]
    RenounceWithSubAccount {},

    #[error("Can't propose Renounced governance, use update_ownership instead")]
    ProposeRenounced {},

    #[error("Can't create account with Renounced governance")]
    InitRenounced {},
}
