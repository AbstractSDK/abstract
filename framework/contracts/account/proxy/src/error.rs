use abstract_core::{objects::ans_host::AnsHostError, AbstractError};
use abstract_sdk::AbstractSdkError;
use cosmwasm_std::{StdError, Uint128};
use cw_asset::AssetError;
use cw_utils::ParseReplyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ProxyError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Asset(#[from] AssetError),

    #[error(transparent)]
    Admin(#[from] ::cw_controllers::AdminError),

    #[error("{0}")]
    Parse(#[from] ParseReplyError),

    #[error("{0}")]
    AnsHostError(#[from] AnsHostError),

    #[error("Module with address {0} is already whitelisted")]
    AlreadyWhitelisted(String),

    #[error("Module with address {0} not found in whitelist")]
    NotWhitelisted(String),

    #[error("Sender is not whitelisted")]
    SenderNotWhitelisted {},

    #[error("Max amount of assets registered")]
    AssetsLimitReached,

    #[error("Max amount of modules registered")]
    ModuleLimitReached,

    #[error("no base asset registered on proxy")]
    MissingBaseAsset,

    #[error("The proposed update resulted in a bad configuration: {0}")]
    BadUpdate(String),

    #[error(
        "Account balance too low, {} requested but it only has {}",
        requested,
        balance
    )]
    Broke {
        balance: Uint128,
        requested: Uint128,
    },

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),
}
