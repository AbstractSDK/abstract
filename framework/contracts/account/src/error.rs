use abstract_sdk::std::objects::module::ModuleInfo;
use abstract_std::{
    objects::{registry::RegistryError, validation::ValidationError},
    AbstractError,
};
use cosmwasm_std::{Instantiate2AddressError, StdError};
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error(transparent)]
    Std(#[from] StdError),

    #[error(transparent)]
    Abstract(#[from] AbstractError),

    #[error(transparent)]
    Validation(#[from] ValidationError),

    #[error(transparent)]
    Ownership(#[from] abstract_std::objects::ownership::GovOwnershipError),

    #[error(transparent)]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error(transparent)]
    RegistryError(#[from] RegistryError),

    #[error("Your account is currently suspended")]
    AccountSuspended {},

    // ** Modules Error ** //
    #[error("Failed to query modules to install: {error}")]
    QueryModulesFailed { error: RegistryError },

    #[error("Module with id: {0} is already installed")]
    ModuleAlreadyInstalled(String),

    #[error("Reinstalls of same version of app or standalone are not allowed")]
    ProhibitedReinstall {},

    #[error("The provided module {0} can't be installed on an Abstract account")]
    ModuleNotInstallable(String),

    #[error("The name of the proposed module can not have length 0.")]
    InvalidModuleName {},

    #[error("The provided module {0} was not found")]
    ModuleNotFound(String),

    #[error("Cannot migrate {} twice", module_id)]
    DuplicateModuleMigration { module_id: String },

    #[error("{0} not upgradable")]
    NotUpgradeable(ModuleInfo),

    #[error("Cannot remove module because {0:?} depend(s) on it.")]
    ModuleHasDependents(Vec<String>),

    #[error("Module {module_id} with version {version} does not fit requirement {comp}, post_migration: {post_migration}")]
    VersionRequirementNotMet {
        module_id: String,
        version: String,
        comp: String,
        post_migration: bool,
    },

    #[error("module {0} is a dependency of {1} and is not installed.")]
    DependencyNotMet(String, String),

    #[error("Max amount of modules registered")]
    ModuleLimitReached {},

    #[error("Module with address {0} is already whitelisted")]
    AlreadyWhitelisted(String),

    #[error("can't remove module that is not whitelisted")]
    NotWhitelisted {},

    // ** Sub Account ** //
    #[error("Removing sub account failed")]
    SubAccountRemovalFailed {},

    #[error("Register of sub account failed")]
    SubAccountRegisterFailed {},

    #[error("Can't renounce account, with active sub account")]
    RenounceWithSubAccount {},

    // ** Other Errors ** //
    #[error("No updates were included")]
    NoUpdates {},

    #[error("invalid configuration action, {}", error)]
    InvalidConfigAction { error: StdError },

    #[error("The provided contract version {0} is lower than the current version {1}")]
    OlderVersion(String, String),

    #[error("Sender is not whitelisted and is not a valid owner")]
    SenderNotWhitelistedOrOwner {},

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("The caller ({caller}) is not the owner account's account ({account}). Only account can create sub-accounts for itself.", )]
    SubAccountCreatorNotAccount { caller: String, account: String },

    #[error("You can't chain admin calls")]
    CantChainAdminCalls {},

    #[error("Abstract Account Address don't match to the Contract address")]
    AbsAccInvalidAddr {
        abstract_account: String,
        contract: String,
    },

    #[error("Abstract Account don't have Authentication")]
    AbsAccNoAuth {},

    #[cfg(feature = "xion")]
    #[error(transparent)]
    AbstractXion(#[from] abstract_xion::error::ContractError),
}
