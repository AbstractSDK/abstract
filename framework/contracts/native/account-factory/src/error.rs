use abstract_core::{objects::version_control::VersionControlError, AbstractError};
use abstract_sdk::AbstractSdkError;
use cosmwasm_std::{Instantiate2AddressError, StdError};
use cw_asset::AssetError;
use cw_controllers::AdminError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountFactoryError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Asset(#[from] AssetError),

    #[error("{0}")]
    Ownership(#[from] cw_ownable::OwnershipError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error("{0}")]
    VersionControlError(#[from] VersionControlError),

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("module {0} is required to be of kind {1}")]
    WrongModuleKind(String, String),

    #[error("Your payment does not match the required payment {0}")]
    WrongAmount(String),

    #[error("No payment received")]
    NoPaymentReceived {},

    #[error("Can not create remote accounts without configured IBC host.")]
    IbcHostNotSet {},

    #[error("A trace must exist of at least one or at most {0} hops but has {1}")]
    InvalidTrace(usize, usize),

    #[error("Sender {0} is not the IBC host {1}")]
    SenderNotIbcHost(String, String),
}
