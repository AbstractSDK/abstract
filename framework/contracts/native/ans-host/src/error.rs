use abstract_sdk::AbstractSdkError;
use abstract_std::AbstractError;
use cosmwasm_std::StdError;
use cw_asset::AssetError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AnsHostError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Asset(#[from] AssetError),

    #[error("{0}")]
    Ownership(#[from] cw_ownable::OwnershipError),

    #[error("You must provide exactly two assets when adding liquidity")]
    NotTwoAssets {},

    #[error("{} is not part of the provided pool", id)]
    NotInPool { id: String },

    #[error("{} assets is not within range [{}-{}]", provided, min, max)]
    InvalidAssetCount {
        min: usize,
        max: usize,
        provided: usize,
    },

    #[error("Dex {} is not registered", dex)]
    UnregisteredDex { dex: String },

    #[error("Asset {} is not registered", asset)]
    UnregisteredAsset { asset: String },

    #[error("Dex {} is already registered", dex)]
    DexAlreadyRegistered { dex: String },
}
