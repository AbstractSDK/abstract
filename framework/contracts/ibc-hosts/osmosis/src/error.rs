use abstract_core::AbstractError;
use abstract_ibc_host::HostError;
use abstract_sdk::AbstractSdkError;
use cosmwasm_std::StdError;
use dex::error::DexError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum OsmoError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    DexError(#[from] DexError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    HostError(#[from] HostError),

    #[error("DEX {0} is not a known dex on this network.")]
    UnknownDex(String),

    #[error("Cw1155 is unsupported.")]
    Cw1155Unsupported,

    #[error("Can't provide liquidity less than two assets")]
    TooFewAssets {},

    #[error("Can't provide liquidity with more than {0} assets")]
    TooManyAssets(u8),

    #[error("Provided asset {0} not in pool with assets {1:?}.")]
    ArgumentMismatch(String, Vec<String>),

    #[error("Pair {0} on DEX {1} does not match with pair address {2}")]
    DexMismatch(String, String, String),
}
