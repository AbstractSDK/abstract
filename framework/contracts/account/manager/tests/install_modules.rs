mod common;

use abstract_core::{
    manager::{
        ExecuteMsg as ManagerMsg, ModuleAddressesResponse, ModuleInstallConfig,
        QueryMsg as ManagerQuery,
    },
    objects::{account::TEST_ACCOUNT_ID, module::ModuleInfo},
};
use abstract_interface::{Abstract, AbstractAccount, VCExecFns};
use abstract_manager::error::ManagerError;
use abstract_testing::prelude::TEST_NAMESPACE;
use common::{
    mock_modules::{adapter_1, deploy_modules, V1},
    *,
};
use cosmwasm_std::Addr;
use cw_orch::{
    deploy::Deploy,
    prelude::{CwOrchExecute, CwOrchQuery, Mock},
    take_storage_snapshot,
};

#[test]
fn cannot_reinstall_module() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let abstr = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&abstr.account_factory)?;

    let AbstractAccount { manager, proxy: _ } = &account;

    abstr
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, TEST_NAMESPACE.to_string())?;

    deploy_modules(&chain);

    manager.execute(
        &ManagerMsg::InstallModules {
            modules: vec![ModuleInstallConfig::new(
                ModuleInfo::from_id(adapter_1::MOCK_ADAPTER_ID, V1.into()).unwrap(),
                None,
            )],
        },
        None,
    )?;

    let err = manager
        .execute(
            &ManagerMsg::InstallModules {
                modules: vec![ModuleInstallConfig::new(
                    ModuleInfo::from_id(adapter_1::MOCK_ADAPTER_ID, V1.into()).unwrap(),
                    None,
                )],
            },
            None,
        )
        .unwrap_err();
    let manager_err: ManagerError = err.downcast().unwrap();
    assert_eq!(
        manager_err,
        ManagerError::ModuleAlreadyInstalled(adapter_1::MOCK_ADAPTER_ID.to_owned())
    );
    Ok(())
}

#[test]
fn adds_module_to_account_modules() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let abstr = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&abstr.account_factory)?;

    let AbstractAccount { manager, proxy: _ } = &account;

    abstr
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, TEST_NAMESPACE.to_string())?;

    deploy_modules(&chain);

    manager.execute(
        &ManagerMsg::InstallModules {
            modules: vec![ModuleInstallConfig::new(
                ModuleInfo::from_id(adapter_1::MOCK_ADAPTER_ID, V1.into()).unwrap(),
                None,
            )],
        },
        None,
    )?;

    let addrs: ModuleAddressesResponse = manager.query(&ManagerQuery::ModuleAddresses {
        ids: vec![adapter_1::MOCK_ADAPTER_ID.to_owned()],
    })?;
    assert_eq!(addrs.modules.len(), 1);
    take_storage_snapshot!(chain, "adds_module_to_account_modules");
    Ok(())
}
