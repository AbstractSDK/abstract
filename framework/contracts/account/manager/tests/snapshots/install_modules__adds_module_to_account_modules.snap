---
source: contracts/account/manager/tests/install_modules.rs
expression: all_storage
---
"abstract:account-factory":
  - - acseq
    - "2"
  - - cfg
    - "{\"version_control_contract\":\"contract1\",\"ans_host_contract\":\"contract0\",\"module_factory_address\":\"contract2\",\"ibc_host\":\"contract5\"}"
  - - contract_info
    - "{\"contract\":\"abstract:account-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - config
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
  - - registered_dexes
    - "[]"
"abstract:ibc-client":
  - - config
    - "{\"version_control\":{\"address\":\"contract1\"},\"ans_host\":{\"address\":\"contract0\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.20.0\"}"
  - - module_data
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.20.0\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - cfg
    - "{\"ans_host\":{\"address\":\"contract0\"},\"account_factory\":\"contract3\",\"version_control\":{\"address\":\"contract1\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-0":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"v\\u0005 W$ÛÒyø<\\\\wÃ\\u0001\\u001Bµg)[3>)¼â\\u001A\\u0002&·ù\\\"m\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-1":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\""
  - - "\u0000\u0007modulestester:mock-adapter1"
    - "\"contract10\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.20.0\"}"
  - - icontext
    - "[[{\"info\":{\"namespace\":\"tester\",\"name\":\"mock-adapter1\",\"version\":{\"version\":\"1.0.0\"}},\"reference\":{\"adapter\":\"contract10\"}},null]]"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - "\u0000{5}config"
    - "{\"version_control_address\":\"contract1\",\"ans_host_address\":\"contract0\"}"
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:proxy-local-0":
  - - "\u0000{5}state"
    - "{\"modules\":[\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - admin
    - "\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.20.0\"}"
"abstract:proxy-local-1":
  - - "\u0000{5}state"
    - "{\"modules\":[\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\",\"contract10\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - admin
    - "\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.20.0\"}"
"abstract:version-control":
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\tmock-app11.0.0"
    - "{\"app\":13}"
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\tmock-app12.0.0"
    - "{\"app\":14}"
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\rmock-adapter11.0.0"
    - "{\"adapter\":\"contract10\"}"
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\rmock-adapter12.0.0"
    - "{\"adapter\":\"contract11\"}"
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\rmock-adapter21.0.0"
    - "{\"adapter\":\"contract12\"}"
  - - "\u0000\u0003lib\u0000\u0006tester\u0000\rmock-adapter22.0.0"
    - "{\"adapter\":\"contract13\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0005proxy0.20.0"
    - "{\"account_base\":6}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0007manager0.20.0"
    - "{\"account_base\":5}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\bans-host0.20.0"
    - "{\"native\":\"contract0\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\nibc-client0.20.0"
    - "{\"native\":\"contract4\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000emodule-factory0.20.0"
    - "{\"native\":\"contract2\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000faccount-factory0.20.0"
    - "{\"native\":\"contract3\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000fversion-control0.20.0"
    - "{\"native\":\"contract1\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "{\"manager\":\"D?Ê\\u0014¬å[#^,¡4\\r\\u000E\\r¡éÓ¯:5Z\\u000EÑüF%¾á\\b\",\"proxy\":\"v\\u0005 W$ÛÒyø<\\\\wÃ\\u0001\\u001Bµg)[3>)¼â\\u001A\\u0002&·ù\\\"m\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "{\"manager\":\"úÃº\\u000F¡f9Ê\\u000E/´K}\\u001D)¨\\u00167ÐvNÚ¡ëã}|d\",\"proxy\":\"ØJ\\\\Ì1¬´B´ÛWn\\\"\\u001B\\u0012`±¿9½\\u0005\\u0002Íºñù¬\"}"
  - - "\u0000\u0005nmspcabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - "\u0000\u0005nmspctester"
    - "{\"trace\":\"local\",\"seq\":1}"
  - - "\u0000\u0007nmspc_a\u0000\u0005local\u0000\u0004\u0000\u0000\u0000\u0000abstract"
    - "8"
  - - "\u0000\u0007nmspc_a\u0000\u0005local\u0000\u0004\u0000\u0000\u0000\u0001tester"
    - "6"
  - - cfg
    - "{\"account_factory_address\":\"contract3\",\"allow_direct_module_registration_and_updates\":true,\"namespace_registration_fee\":null}"
  - - contract_info
    - "{\"contract\":\"abstract:version-control\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"tester:mock-adapter1":
  - - base_state
    - "{\"version_control\":{\"address\":\"contract1\"},\"ans_host\":{\"address\":\"contract0\"}}"
  - - contract_info
    - "{\"contract\":\"tester:mock-adapter1\",\"version\":\"2.0.0\"}"
  - - module_data
    - "{\"module\":\"tester:mock-adapter1\",\"version\":\"2.0.0\",\"dependencies\":[],\"metadata\":null}"
"tester:mock-adapter2":
  - - base_state
    - "{\"version_control\":{\"address\":\"contract1\"},\"ans_host\":{\"address\":\"contract0\"}}"
  - - contract_info
    - "{\"contract\":\"tester:mock-adapter2\",\"version\":\"2.0.0\"}"
  - - module_data
    - "{\"module\":\"tester:mock-adapter2\",\"version\":\"2.0.0\",\"dependencies\":[],\"metadata\":null}"

